cpp/include/cudf/io/functions.hpp

```diff
@@ -428,11 +433,149 @@ struct read_parquet_args {
   /// Cast timestamp columns to a specific type
   data_type timestamp_type{type_id::EMPTY};
 
+  explicit read_parquet_args(source_info const& src) : source(src) {}
+
+  friend read_parquet_args_builder;
+
+ public:
   explicit read_parquet_args() = default;
 
-  explicit read_parquet_args(source_info const& src) : source(src) {}
+  /**
+   * @brief create read_parquet_args_builder which will build read_parquet_args
+   *
+   * @param src source information used to read parquet file
+   * @returns read_parquet_args_builder
+   */
+  static read_parquet_args_builder build(source_info const& src);
+
+  /**
+   * @brief Returns source info
+   */
+  source_info get_source_info() const { return source; }
+
+  /**
+   * @brief Returns column names
+   */
+  std::vector<std::string> column_names() const { return columns; }
+
+  /**
+   * @brief Returns row groups
+   */
+  std::vector<std::vector<size_type>> get_row_groups() const { return row_groups; }
+
+  /**
+   * @brief Returns number of rows to skip
+   */
+  size_type rows_to_skip() const { return skip_rows; }
+
+  /**
+   * @brief Returns row groups
+   */
+  size_type get_num_rows() const { return num_rows; }
+
+  /**
+   * @brief Returns strings_to_categorical
+   */
+  bool is_strings_to_categorical() const { return strings_to_categorical; }
+
+  /**
+   * @brief Returns use_pandas_metadata
+   */
+  bool utilize_pandas_metadata() const { return use_pandas_metadata; }
+
+  /**
+   * @brief Returns timestamp type
+   */
+  data_type get_timestamp_type() const { return timestamp_type; }
 };
 
+class read_parquet_args_builder {
+  read_parquet_args args;
+
+ public:
+  explicit read_parquet_args_builder() = default;
+
+  explicit read_parquet_args_builder(source_info const& src) : args(src) {}
+
+  /**
+   * @brief Set column names which needs to be read
+   */
+  read_parquet_args_builder& with_column_names(std::vector<std::string> column_names)
+  {
+    args.columns = column_names;
+    return *this;
+  }
+
+  /**
+   * @brief Set row groups
+   */
+  read_parquet_args_builder& with_row_groups(std::vector<std::vector<size_type>> row_groups)
+  {
+    args.row_groups = row_groups;
```

Comment:
Passing by value and assigning like this will cause move+copy or copy+copy. We can improve this:
```suggestion
    args.row_groups = std::move(row_groups);
```
with this, the calls becomes two moves or a copy+move. [Detailed explanation](https://dzone.com/articles/c-sink-parameter-passing#:~:text=If%20a%20parameter%20is%20movable,and%20copy%20from%20anything%20else.).

This applies to all parameters that are expensive to copy.